### ISSUE TITLE
Scraping takes a **long** time

### ISSUE THREAD

**Issue by kcmcleod** [HW-SWeL/Scraper#16]
From 3.30pm yesterday until 8.30am today I tried to scrape Chembl. I used:
* 4 threads
* each thread had 1 second wait after it completed
* testing suggested each thread would take about 1.5 seconds (including the delay)

In reality, we scraped 7117 pages. Thus only achieved 418 an hour or 7 a minute.

There is a big gap between 7 per minute and each scrape taking 1.5 seconds! Even a single thread should download far more than 7 per minute.

There are 1879206 marked up pages in Chembl (assuming every compound is marked up). Assuming the same rate as yesterday it will take 4,496 hours (ie 188 days) to scrape it.

**Comment by kcmcleod** [comment]
Trying to discover what happened:
* I wondered if i had run out of disk space, thus scrapes didn't save? No, I have 150GB spare.
* Run a short experiment (only 30 scrapes):
** 1st scrape takes about 4.5s
** subsequent scrapes drop to between 1.2 and 1.5s; average works out at 1.3s

Looking at the log, we have a lot of errors! So many the log file overwrote itself. It seems that not all the compound pages in Chembl are marked up, eg, https://www.ebi.ac.uk/chembl/compound_report_card/CHEMBL1163153.

In summary the result above are invalid! Need to try again with pages that I know for sure work :(

**Comment by kcmcleod** [comment]
New test; scraping 100 pages (that I *know* have markup) with:
* 2 threads takes 78s
* 4 threads takes 45s

So much faster than 7 per minute.

**Comment by kcmcleod** [comment]
On my 8 core machine, 4 threads seems to use 4 cores at about 20-45% each.
Room to use more threads?

**Comment by kcmcleod** [comment]
Next experiment; same machine with 4 threads:
* 3997 pages took 29 minutes
* 49 pages could not be scraped

**Comment by kcmcleod** [comment]
Next experiment; same machine, 4 threads:
* 11993 pages took 54 minutes
* 146 pages could not be scraped

**Comment by kcmcleod** [comment]
Next experiment; same machine, but 6 threads:
* 23999 pages took 55 minutes
* 220 pages could not be scraped

**Comment by kcmcleod** [comment]
Next experiment; same machine, but 8 threads & 2G for VM:
* now using all 8 cores, but still not a lot. 4 cores sitting around 45-70% and the other 4 stay at around 10% 
*

**Comment by kcmcleod** [comment]
Following night I managed over 300000, so not sure what happened on night 1.
